tracing-flame/src/lib.rs

```diff
@@ -0,0 +1,277 @@
+//!
+//! ## Usage
+//!
+//! This crate is meant to be used as a two step process. First, you capture a
+//! textual representation of the spans that are entered and exited with the
+//! [`FlameLayer`] then you feed these into `inferno-flamegraph` to generate the
+//! flamegraph/flamechart image.
+//!
+//! ## Layer Setup
+//!
+//! ```rust
+//! fn setup_global_subscriber() {
+//!     use tracing_flame::FlameLayer;
+//!     use tracing_subscriber::{registry::Registry, prelude::*};
+//!
+//!     let subscriber = Registry::default()
+//!         .with(FlameLayer::write_to_file("./tracing2.folded").unwrap());
+//!
+//!     tracing::subscriber::set_global_default(subscriber).expect("Could not set global default");
+//! }
+//!
+//! // your code here ..
+//! ```
+//!
+//! Alternatively you can provide any type that implements `std::io::Write` to
+//! `FlameLayer::new`.
+//!
+//! ## Generating The Image
+//!
+//! To convert your text representation to a visual one first install inferno
+//!
+//! ```console
+//! cargo install inferno
+//! ```
+//!
+//! Then pass the file your error layer created into `inferno-flamegraph`
+//!
+//! ```console
+//! # flamegraph
+//! cat tracing.folded | inferno-flamegraph > tracing-flamegraph.svg
+//!
+//! #flamechart
+//! cat tracing.folded | inferno-flamegraph --flamechart > tracing-flamechart.svg
+//! ```
+//!
+//! ## Differences between `flamegraph`s and `flamechart`s
+//!
+//! By default `inferno-flamegraph` creates `flamegraph`s. Flamegraphs is that
+//! they collapse identical stack frames and sort them based on their names.
+//! This is great for multithreaded programs and long running programs where the
+//! same frames happen many times for short durations because it simplifies the
+//! graph and gives you a broad idea of overall time spent in each part of the
+//! stack.
+//!
+//! However it is sometimes desirable to preserve the exact ordering of events
+//! as they were emitted by `tracing-flame`, so you can see exactly when each
+//! span is entered relative to the other and get an accurate visual trace of
+//! the execution of your program. To get this style of representation you
+//! should instead generate a `flamechart`, which does not sort or collapse
+//! identical stack frames.
+//!
+//! [`tracing`]: https://docs.rs/tracing
+//! [`inferno`]: https://docs.rs/inferno
+//! [`FlameLayer`]: struct.FlameLayer.html
+mod error;
+
+use error::Error;
+use std::fmt;
+use std::fmt::Write as _;
+use std::fs::File;
+use std::io::BufWriter;
+use std::io::Write;
+use std::marker::PhantomData;
+use std::path::Path;
+use std::sync::Mutex;
+use std::time::{Duration, Instant};
+use tracing::span::Attributes;
+use tracing::Id;
+use tracing::Subscriber;
+use tracing_subscriber::layer::Context;
+use tracing_subscriber::registry::LookupSpan;
+use tracing_subscriber::registry::SpanRef;
+use tracing_subscriber::Layer;
+
+pub struct FlameLayer<S, W>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+    W: Write + 'static,
```

Comment:
WDYT about using the [`MakeWriter` trait](https://docs.rs/tracing-subscriber/0.2.3/tracing_subscriber/fmt/trait.MakeWriter.html) from `tracing_subscriber` here? Since `io::stdout` already has locking built in, if we wanted to use stdout here, we would be adding an unnecessary layer of locking with the mutex, and storing the stdout handle in the mutex would prevent anything else from being able to write to stdout.

We also want to eventually implement the dedicated logging thread stuff using the `MakeWriter` interface, so if we used that here, we could layer that in so that the flamegraph is written on a dedicated thread without needing to do any extra work.

Reply:
sounds like a plan